server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java

```diff
@@ -397,6 +397,32 @@ record StartRecoveryRequestToSend(StartRecoveryRequest startRecoveryRequest, Str
                     }
                     indexShard.recoverLocallyUpToGlobalCheckpoint(ActionListener.assertOnce(l));
                 })
+                // peer recovery can consume a lot of disk space, so it's worth cleaning up locally ahead of the attempt
+                // operation runs only if the previous operation succeeded, and returns the previous operation's result.
+                // Failures at this stage aren't fatal, we can attempt to recover and then clean up again at the end. #104473
+                .andThenApply(startingSeqNo -> {
+                    Store store = indexShard.store();
+                    store.incRef();
+                    try {
+                        store.cleanupAndVerify("cleanup before peer recovery", indexShard.snapshotStoreMetadata());
+                    } catch (IOException e) {
```

Comment:
I think I might prefer to restrict this `catch` just to exceptions from `indexShard.snapshotStoreMetadata()` rather than from `cleanupAndVerify()` too. If `cleanupAndVerify()` fails then that seems like it maybe should be fatal?

Also could we use `org.elasticsearch.index.store.Store.MetadataSnapshot#EMPTY` to represent the snapshot if we cannot get a proper `MetadataSnapshot`? Any file re-use during recovery is going to require a valid `MetadataSnapshot`, so failing here means that the directory contents are of no further use.

Reply:
Tim made a similar comment, to which I replied above. There is some logic in `cleanFiles` that inspects the exception generated by the block containing `cleanupAndVerify` and does a deeper clean in some cases. Looking at it again, I think it's not just an optimization. We'd probably need to replicate that to avoid having to worry about corrupt directories that we can never make forward progress on because we never get a chance to deep clean them.

Is it always true that the files are useless if we can't get a metadata snapshot? I had been thinking that for instance we might have transferred a bunch of data files previously but not the root segment file. In that case, after transferring the segment file we could reuse the data files already transferred?

Reply:
> Is it always true that the files are useless if we can't get a metadata snapshot?

Yes I think so, we treat a failure as if the store is empty, see how we compute `metadataSnapshot` in `org.elasticsearch.indices.recovery.PeerRecoveryTargetService#getStartRecoveryRequest` and then use this data (alone) to compute the collection of files to recover in the implementations of `org.elasticsearch.indices.recovery.plan.RecoveryPlannerService#computeRecoveryPlan`.

I mean you're right, in principle we could work harder to determine which exact files we do/don't have and avoid a little extra data transfer sometimes. But we don't: this is already on a failure path, and not a particularly well-trodden one, so there's no need for heroic optimizations.

Reply:
I thought a bit more about this and am inclined to use `Store.MetadataSnapshot#EMPTY` when `snapshotStoreMetadata()` throws. Interestingly, when I tried it, I found that `store.cleanupAndVerify` appears to throw on an empty directory, [here](https://github.com/elastic/elasticsearch/blob/8efd08b019b9160f5e703520c6ce3b6a9f92cfbd/server/src/main/java/org/elasticsearch/index/store/Store.java#L699):

```
...
Caused by: java.io.FileNotFoundException: no segments* file found in store(ByteSizeCachingDirectory(ElasticsearchMockDirectoryWrapper(NIOFSDirectory@/Users/brendan/dev/elasticsearch/server/build/testrun/internalClusterTest/temp/org.elasticsearch.recovery.TruncatedRecoveryIT_31340047D7FA1434-001/tempDir-002/node_t4/indices/QuuB7D0wSki7rj5uryRXWQ/4/index lockFactory=org.apache.lucene.store.NativeFSLockFactory@b7a4d9))): files: [write.lock]
	at org.apache.lucene.index.SegmentInfos$FindSegmentsFile.run(SegmentInfos.java:801)
	at org.apache.lucene.index.SegmentInfos$FindSegmentsFile.run(SegmentInfos.java:757)
	at org.apache.lucene.index.SegmentInfos.readLatestCommit(SegmentInfos.java:535)
	at org.apache.lucene.index.SegmentInfos.readLatestCommit(SegmentInfos.java:519)
	at org.elasticsearch.common.lucene.Lucene.readSegmentInfos(Lucene.java:116)
	at org.elasticsearch.index.store.Store.readSegmentsInfo(Store.java:224)
	at org.elasticsearch.index.store.Store$MetadataSnapshot.loadFromIndexCommit(Store.java:822)
	at org.elasticsearch.index.store.Store.getMetadata(Store.java:298)
	at org.elasticsearch.index.store.Store.getMetadata(Store.java:263)
	at org.elasticsearch.index.store.Store.cleanupAndVerify(Store.java:699)
...
```

Given the name of the variable being assigned here is `metadataOrEmpty` it seems like it might be a bug, and we just haven't tested against an empty snapshot. If you agree, I'll attempt to fix that and go with this approach of falling back to `Empty`.

Reply:
https://github.com/elastic/elasticsearch/pull/116059 adds support for empty stores to `Store#cleanupAndVerify`

Reply:
after making `cleanupAndVerify` handle an empty snapshot, I've taken your suggestion and treated the directory as having an empty snapshot if we fail to load one.